src/EdgeConProblem/EdgeConReduction.c

```diff
@@ -8,36 +8,290 @@
+
+/**
+ * FIXME: is the DM2 report h start at 1 but it should start at 2 because of the h-1.
+ *
+ * Builds the formula ensuring the constraint:
+ *
+ *   "For two homogeneous components, if X_@p j1 is at level h then X_@p j2
+ *    is at level h - 1."
+ *
+ * @param ctx is the current reduction context.
+ * @param j1 is the number of the first homogeneous component.
+ * @param j2 is the number of the second homogeneous component.
+ *
+ * @return the Z3 ast corresponding to the formula.
+ */
+static Z3_ast build_phi_7(const g_context_s *ctx, const int j1, const int j2);
+
+/**
+ * Builds the formula ensuring the constraint:
+ *
+ *   "For any two homogeneous components, if X_j1 is a parent of X_j2, then the
+ *    conditions of phi_6 and phi_7 are satisfied."
+ *
+ * @param ctx is the current reduction context.
+ *
+ * @return the Z3 ast corresponding to the formula.
+ */
+static Z3_ast build_phi_8(const g_context_s *ctx);
+
+static g_context_s* init_g_context(Z3_context z3_ctx, EdgeConGraph graph, int cost);
+
+/**
+ * Checks if the edge (@p n1, @p n2) is the @p i -th translator.
+ *
+ * @param ctx is the solver context.
+ * @param model is the solver model.
+ * @param graph is a EdgeConGraph.
+ * @param n1 is the first node of the edge.
+ * @param n2 is the second node of the edge.
+ * @param i is number of the the translator.
+ *
+ * @return true if the edge (@p n1, @p n2) is the @p i -th translator of @p
+ * graph, otherwise false.
+ */
+static bool is_the_ith_translator(
+    const Z3_context ctx,
+    const Z3_model model,
+    const EdgeConGraph graph,
+    const int n1,
+    const int n2,
+    const int i
+);
+
+Z3_ast getVariableIsIthTranslator(Z3_context ctx, int node1, int node2, int number) {
     char name[40];
-    if (node1 < node2)
-        snprintf(name, 40, "[(%d,%d),%d]", node1, node2, number);
-    else
-        snprintf(name, 40, "[(%d,%d),%d]", node2, node1, number);
+
+    if (node1 < node2) {
+        snprintf(name, 40, "x_[(%d,%d),%d]", node1, node2, number);
+    }
+    else {
+        snprintf(name, 40, "x_[(%d,%d),%d]", node2, node1, number);
+    }
+
     return mk_bool_var(ctx, name);
 }
 
-Z3_ast getVariableParent(Z3_context ctx, int child, int parent)
-{
+Z3_ast getVariableParent(Z3_context ctx, int child, int parent) {
     char name[40];
+
     snprintf(name, 40, "p_[%d,%d]", child, parent);
+
     return mk_bool_var(ctx, name);
 }
 
-Z3_ast getVariableLevelInSpanningTree(Z3_context ctx, int level, int component)
-{
+Z3_ast getVariableLevelInSpanningTree(Z3_context ctx, int level, int component) {
     char name[40];
+
     snprintf(name, 40, "l_[%d,%d]", component, level);
+
     return mk_bool_var(ctx, name);
 }
 
-Z3_ast EdgeConReduction(Z3_context ctx, EdgeConGraph edgeGraph, int cost)
-{
-    return Z3_mk_false(ctx);
+Z3_ast EdgeConReduction(Z3_context z3_ctx, EdgeConGraph edgeGraph, int cost) {
+    g_context_s *ctx;
+
+    ctx = init_g_context(z3_ctx, edgeGraph, cost);
+
+    return(
+        OR(5)
```

Comment:
Ca devrait être AND :
```suggestion
        AND(5)
```